### PR TITLE
Add client deploy support for RE projects

### PR DESCRIPTION
**Pull request by feluelle**
## Description

> Describe the purpose of this pull request.

This pull request introduces a new `--client` flag to the `astro deploy` command, enabling users to push a client image directly to a remote registry. The implementation includes a new code path for handling client image deployment, robust error handling, and comprehensive unit tests for both the CLI and deployment logic. Additionally, the Docker login process is refactored to support OAuth-based registries and is made more testable.

**Client image deployment feature:**

* Added a `--client` flag to the `astro deploy` CLI command, allowing users to push a client image to a remote registry. The flag sets the new `ClientDeploy` field in `InputDeploy`, which triggers the new deployment logic. [[1]](diffhunk://#diff-e6ab40766b237bea11c753ce8f4f94ea143ee20c0794b8d41cfb6f1c39bbe24dR44) [[2]](diffhunk://#diff-e6ab40766b237bea11c753ce8f4f94ea143ee20c0794b8d41cfb6f1c39bbe24dR79) [[3]](diffhunk://#diff-e6ab40766b237bea11c753ce8f4f94ea143ee20c0794b8d41cfb6f1c39bbe24dR160) [[4]](diffhunk://#diff-05e6dbfa75ff6117dd1c27f366e81a20392fd867e25cbeb5693aa1224bdcf7ffR113) [[5]](diffhunk://#diff-05e6dbfa75ff6117dd1c27f366e81a20392fd867e25cbeb5693aa1224bdcf7ffR196-R199)
* Implemented the `deployClientImage` function in `cloud/deploy/deploy.go`, which builds and pushes the client image, handles custom image names/tags, authenticates with the registry, and provides detailed error messages.
* Added CLI and deployment tests to verify that the `--client` flag sets the correct fields, triggers the right code path, and handles edge cases (e.g., missing registry, build/push failures, custom image names). [[1]](diffhunk://#diff-5f76d8147d8b9af21c994836aabeea86cc637a7cd974a44e7aec11354752be4bR69-R102) [[2]](diffhunk://#diff-6b4bd2cc84b9d137148ec88680520827b20bd6c7ad7207bafd358eeb56cda959R1300-R1493)

**Docker login improvements:**

* Refactored the Docker login logic to use a bash command for authentication, supporting OAuth-based registries and stripping Bearer prefixes from tokens. The login function is now assigned to a variable (`DockerLogin`) for easier mocking in tests. [[1]](diffhunk://#diff-0c6d2535c584831a8191a01fef696ac8f61ec30731e4ec95373e29035c1f44a7R7-R9) [[2]](diffhunk://#diff-0c6d2535c584831a8191a01fef696ac8f61ec30731e4ec95373e29035c1f44a7R78-R102)
* Added comprehensive unit tests for the new Docker login logic, covering successful logins, Bearer token handling, empty credentials, command failures, and runtime errors. [[1]](diffhunk://#diff-a80a9fb770e0734024d6b483165d48a475317a5e2a7c9a9de0b529400b9a6c88R5-R9) [[2]](diffhunk://#diff-a80a9fb770e0734024d6b483165d48a475317a5e2a7c9a9de0b529400b9a6c88R51-R151)

These changes make the deployment process more flexible for advanced use cases and improve the reliability and testability of Docker registry authentication.

## 🎟 Issue(s)

closes: https://github.com/astronomer/astro-cli/issues/1929

## 🧪 Functional Testing

> List the functional testing steps to confirm this feature or fix.

## 📸 Screenshots

> Add screenshots to illustrate the validity of these changes.

## 📋 Checklist

- [x] Rebased from the main (or release if patching) branch (before testing)
- [x] Ran `make test` before taking out of draft
- [x] Ran `make lint` before taking out of draft
- [x] Added/updated applicable tests
- [x] Tested against [Astro-API](https://github.com/astronomer/astro/) (if necessary).
- [ ] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

File: airflow/docker_registry.go [hunk 1]
```diff
@@ -73,3 +75,28 @@ func (d *DockerRegistry) Login(username, token string) error {
 	}
 	return nil
 }
+
+// dockerLogin performs docker login using bash command instead of Docker API
+// This is useful for OAuth-based registries like Astronomer that require special authentication flows
```
Comment (neel-astro) [inline airflow/docker_registry.go]:
```suggestion
// This is useful for OAuth-based registries that require special authentication flows
```

File: airflow/docker_registry.go [hunk 2]
```diff
@@ -73,3 +75,28 @@ func (d *DockerRegistry) Login(username, token string) error {
...
+		if err != nil {
+			return fmt.Errorf("docker login failed: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// DockerLogin is a testable variable that holds the docker login function
+var DockerLogin = dockerLogin
```
Comment (neel-astro) [inline airflow/docker_registry.go]: nit: maybe just move it to the top, for ease of reading

File: cloud/deploy/deploy.go [hunk 3]
```diff
@@ -801,3 +807,75 @@ func WarnIfNonLatestVersion(version string, httpClient *httputil.HTTPClient) {
...
+
+// deployClientImage handles the client deploy functionality
+func deployClientImage(deployInput InputDeploy, c *config.Context) error { //nolint:gocritic
+	fmt.Println("Deploying client image...")
+
+	// Get the remote client registry endpoint from config
+	registryEndpoint := config.CFG.RemoteClientRegistry.GetString()
+	if registryEndpoint == "" {
+		return errors.New("remote client registry is not configured. Please run 'astro dev init' with remote execution enabled first, or configure it using 'astro config set remote.client_registry <endpoint>'")
+	}
```
Comment (neel-astro) [inline cloud/deploy/deploy.go:828]:
```suggestion
	if registryEndpoint == "" {
		return errors.New("remote client registry is not configured. Please run 'astro config set remote.client_registry <endpoint>' to configure the registry")
	}
```

nit: At this point, users will already have the local project going, so making them create a new project again might be counterintuitive, so simply stating `astro config set` command might be the right thing

File: cloud/deploy/deploy.go [hunk 4]
```diff
@@ -801,3 +807,75 @@ func WarnIfNonLatestVersion(version string, httpClient *httputil.HTTPClient) {
...
+	var imageTag string
+	if deployInput.ImageName != "" {
+		// Extract tag from custom image name if provided
+		parts := strings.Split(deployInput.ImageName, ":")
+		if len(parts) == 2 {
+			imageTag = parts[1]
+		} else {
+			imageTag = "latest"
+		}
+	} else {
```
Comment (neel-astro) [inline cloud/deploy/deploy.go]: Open question: I'm unsure if copying the tag is the right approach. For regular image deploy, we continue to use our own tagging logic, which may provide consistency.

File: cloud/deploy/deploy.go [hunk 5]
```diff
@@ -801,3 +807,75 @@ func WarnIfNonLatestVersion(version string, httpClient *httputil.HTTPClient) {
...
+	remoteImage := fmt.Sprintf("%s:%s", registryEndpoint, imageTag)
+
+	// Create an image handler for building and pushing
+	localImageName := airflow.ImageName("agent-client", "latest")
+	imageHandler := airflowImageHandler(localImageName)
+
+	if deployInput.ImageName != "" && strings.Contains(deployInput.ImageName, ":") {
+		// If a full image name with tag is provided, use it directly
+		fmt.Println("Using provided image:", deployInput.ImageName)
+		err := imageHandler.TagLocalImage(deployInput.ImageName)
```
Comment (neel-astro) [inline cloud/deploy/deploy.go:843]: I think over here the imageHandler needs to be `airflowImageHandler(remoteImage)`, so that the local image is tagged as a remote image

File: cloud/deploy/deploy.go [hunk 6]
```diff
@@ -801,3 +807,75 @@ func WarnIfNonLatestVersion(version string, httpClient *httputil.HTTPClient) {
...
+		if err != nil {
+			return fmt.Errorf("failed to tag local image: %w", err)
+		}
+	} else {
+		// Authenticate with the base image registry before building
+		// This is needed because Dockerfile.client uses base images from a private registry
+		baseImageRegistry := "images.astronomer.cloud"
```
Comment (neel-astro) [inline cloud/deploy/deploy.go]: Maybe move it to configuration, so that we can use dev for development

File: cloud/deploy/deploy.go [hunk 7]
```diff
@@ -801,3 +807,75 @@ func WarnIfNonLatestVersion(version string, httpClient *httputil.HTTPClient) {
...
+		err := airflow.DockerLogin(baseImageRegistry, registryUsername, c.Token)
+		if err != nil {
+			return fmt.Errorf("failed to authenticate with registry %s: %w", baseImageRegistry, err)
+		}
+
+		// Build the client image from the current directory
+		fmt.Println("Building client image...")
+		buildConfig := types.ImageBuildConfig{
+			Path:            deployInput.Path,
+			TargetPlatforms: deployImagePlatformSupport,
```
Comment (neel-astro) [inline cloud/deploy/deploy.go]: I think we should be open to allowing image build for other platforms via the input flag only when building client images

**Review by neel-astro — COMMENTED** [review]
Left a couple of comments, but otherwise the change is looking solid, thanks for putting this all together

File: cloud/deploy/deploy.go (hunk 4 above)
Comment (feluelle) [reply cloud/deploy/deploy.go]: From the linked issue:
> By default, we would continue to use the existing deploy-<timestamp> tagging mechanism. If users want to have a custom tag, they could build and tag on their own and then use astro deploy --client --image-name <image-name> to push the image to the registry.

Why should we not give them a way to overwrite the tag? The default is the `deploy-<timestamp>`.

File: cmd/cloud/deploy.go [hunk 8]
```diff
@@ -75,6 +77,8 @@ func NewDeployCmd() *cobra.Command {
 	cmd.Flags().StringVarP(&deployDescription, "description", "", "", "Add a description for more context on this deploy")
 	cmd.Flags().StringSliceVar(&buildSecrets, "build-secrets", []string{}, "Mimics docker build --secret flag. See https://docs.docker.com/build/building/secrets/ for more information. Example input id=mysecret,src=secrets.txt")
 	cmd.Flags().BoolVar(&forceUpgradeToAF3, "force-upgrade-to-af3", false, "Force allow upgrade from Airflow 2 to Airflow 3")
+	cmd.Flags().BoolVar(&clientDeploy, "client", false, "Push the client image to the remote registry")
+	cmd.Flags().StringVar(&platform, "platform", "", "Target platform for client image build (e.g., linux/amd64,linux/arm64) - only works with --client flag")
```
Comment (neel-astro) [inline cmd/cloud/deploy.go]: nit: let's also document the default behaviour as well, which is to build for host machine platform

File: cmd/cloud/remote.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  $ astro remote deploy --build-secrets id=mysecret,src=secrets.txt
+`
+)
+
+// newRemoteRootCmd creates the root command for remote operations
+func newRemoteRootCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "remote",
+		Short: "Manage remote deployments and images",
+		Long:  "Commands for interacting with remote registries and deploying client images",
```
Comment (neel-astro) [inline cmd/cloud/remote.go:42]:
```suggestion
		Long:  "Commands for interacting with remote deployments, registries and deploying client images",
```

File: cmd/cloud/remote.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	cmd.AddCommand(newRemoteDeployCmd())
+	return cmd
+}
+
+// newRemoteDeployCmd creates the remote deploy command
+func newRemoteDeployCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "deploy",
+		Short:   "Deploy a client image to the remote registry",
+		Long:    "Build and deploy a client image to the configured remote registry. This command assumes you have already authenticated with the registry using 'docker login'.",
```
Comment (neel-astro) [inline cmd/cloud/remote.go]:
```suggestion
		Long:    "Build and deploy a client image to the configured remote registry. This command assumes you have already authenticated with the registry.",
```
nit: not calling out Docker here explicitly to not confuse users that are using podman or other container engines

File: cmd/cloud/remote.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+		PreRunE: utils.EnsureProjectDir,
+		RunE:    remoteDeploy,
+		Example: remoteDeployExample,
+	}
+
+	cmd.Flags().StringVar(&remotePlatform, "platform", "", "Target platform for client image build (e.g., linux/amd64,linux/arm64). Defaults to host machine platform")
+	cmd.Flags().StringVarP(&remoteImageName, "image-name", "i", "", "Name of a custom image to deploy, or image name with custom tag")
```
Comment (neel-astro) [inline cmd/cloud/remote.go]:
```suggestion
	cmd.Flags().StringVarP(&remoteImageName, "image-name", "i", "", "Name of a custom image to deploy, or image name with custom tag. The image should be present on the local machine.")
```

File: software/deploy/deploy.go [hunk 12]
```diff
@@ -235,6 +236,7 @@ func pushDockerImage(byoRegistryEnabled bool, deploymentInfo *houston.Deployment
 		}
 	}
 	useShaAsTag := config.CFG.ShaAsTag.GetBool()
+	fmt.Println("Pushing image to Astronomer registry")
```
Comment (neel-astro) [inline software/deploy/deploy.go]:
```suggestion
	fmt.Printf("Pushing image to %s registry\n", registry)
```
nit: Since we are here correcting it, might as well do it the right way. APC can have a use case for bring your own registry, hence simply calling it `Astronomer` registry might be inaccurate.

File: cmd/cloud/remote.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+  $ astro remote deploy --build-secrets id=mysecret,src=secrets.txt
+`
+)
+
+// newRemoteRootCmd creates the root command for remote operations
+func newRemoteRootCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "remote",
+		Short: "Manage remote deployments and images",
```
Comment (feluelle) [inline cmd/cloud/remote.go]:
```suggestion
		Short: "Manage remote deploys and images",
```

File: cmd/cloud/remote.go (hunk 9 above)
Comment (feluelle) [reply cmd/cloud/remote.go:42]: Deployment mgmt can be done exclusively through the `astro deployment` command.

File: software/deploy/deploy.go (hunk 12 above)
Comment (feluelle) [reply software/deploy/deploy.go]:
```suggestion
	fmt.Println("Pushing image to configured registry")
```
